### ISSUE TITLE
Issue with packaging using custom build

### ISSUE THREAD

**Issue by 1runeberg** [ValveSoftware/steamvr_unreal_plugin#128]
Thanks for the reply Rune. I've reviewed my plugin version which I did get from the marketplace but comparing the version number to the latest GitHub version i'm seeing that they are the same revision? I'm running a source build of 4.22.3 that I've cherry picked fixes for and outside of being able to get this plugin to work its the most stable and (bug free) version of the engine I've had since 4.19 hotfix. I copied this marketplace (4.22) plugin into my engine plugins/marketplace directory similar to what it would do for the binary engine instead of placing in my project (electric nodes is installed this way and I've never had issues before) but maybe I need to place it in my project folder as instructed on the GitHub docs? Another question is that when I package using dev I seem to have my bindings. The button mappings work and the files are present in windowsnoeditor/projectname/config/steamvrbindings and checking them does indeed show the correct json structure. I simply do not have tracking. I've updated the VRE and OpenVRExpansion plugin that Mordentral updated in Feb to remove the input refs from. It doesn't seem i'm having the same issues as described in #112. I also ran into the issues listed with having both the oculus and steamvr plugin active but found out that disabling the oculus plugin fixed the confusion the engine was having with sdk/steamvr inputs and binding corruption early on. VRE, OpenVRE, fingersolver, electricnodes, steamvr and steamvr input are my only plugins in the project. Any advice would be greatly appreciated.
![inputplugin](https://user-images.githubusercontent.com/45907836/75750223-5a479f00-5ce9-11ea-86ed-daf72efd0e5f.PNG)

I've been packaging without issue using legacy steamvr and oculus sdk inputs but due to knowing that 4.24 has this input binding method going forward and the only way to get true index support is using this plugin (due to missing keys from the legacy steamvr input) to explore all of the options I'd like to incorporate for input across all platforms I feel that this is the only path forward both from a project migration readiness standpoint and from a full index support standpoint.

_Originally posted by @uno1982 in https://github.com/ValveSoftware/steamvr_unreal_plugin/issues/50#issuecomment-593799186_

**Comment by 1runeberg** [comment]
Hi @uno1982 - if you're using the source build and you also have the marketplace added to the engine (via the Launcher UI), then you need to put the plugin in hte project folder to ensure you are using the correct version.

Building the source version for the engine, you need to build it like so:
"C:\Program Files\Epic Games\UE_4.22\Engine\Build\BatchFiles\RunUAT.bat" BuildPlugin -Plugin=C:\GitHub\SteamVRTemplate\Plugins\steamvr_unreal_plugin\SteamVRInput\SteamVRInput.uplugin -Package=D:\SteamVRPlugin-Builds\Temp -Rocket -TargetPlatforms=Win64 -nocompile

BUT make sure you to then remove the Marketplace version from the LauncherUI.

The simplest is just using it as a project plugin.

Also, make sure you are building for 64-bit , not 32-bit and failing that, enable logging for your release build and see check the logs if the plugin is activating and reporting to the SteamVR Runtime properly (check entries with [STEAMVR] on it.

**Comment by uno1982** [comment]
Thanks for opening this Rune. I’ve included my log  within a comment on the original tracking issue post. I’ve also been told that 4.24 is having the same issue with the newest plugin from a colleague. I greatly appreciate you opening and looking into this.

**Comment by 1runeberg** [comment]
Hi - Thanks for the log snippet, can you send me a complete UE4 log please FROM the build so I can take a better look? Also, if you can send a SteamVR System Report:
![image](https://user-images.githubusercontent.com/17371351/76562568-bbe1e780-64f9-11ea-875d-548ed83085a6.png)

You can send them to runeb@valvesoftware.com

For 4.24, please note the Issue #112 for needing to copy the SteamVRBindings folder manually (they must be under the Game folder) -- note the engine may copy some bindings there if your project is C++ based (it's own SteamVRBindings) but they could potentially be the wrong bindings. Epic is fixing this Engine issue in 4.26.

**Comment by uno1982** [comment]
Rune,
Just wanted to follow up that I've emailed both the files you requested. The full log from the packaged build and the steamvr system report. I'm on version 1.10.22 (01583280986) so that anyone tracking this will have that information.

**Comment by N1LSS0N** [comment]
hey, im having the same issues, im on 4.24, tracking works when in editor, but not in a packaged one,  Oculus SDK disabled and only running throught steamvr cause thats one of the first things i tried. Also inputs works... Im not sure what to do really.  [ALSO USING LAUNCHER ENGINE VERSION 4.24.3]

[EDIT: i fixed mine, basically it now worked when updating my steamworks settings and readded the manifest there.]

**Comment by uno1982** [comment]
Rune, I got curious so I loaded up both my vive pro with index controllers and oculus to test my packaged build. I'll confirm that for 4.22.3 Both Oculus and Steamvr devices fail to track (with oculus sdk disabled), both do however have the correct bindings and input events are firing correctly. I tried loading up multiple times into each just to be sure. My oculus only tracks if I leave that plugin enabled as that module takes priority over steamvr. I currently do not plan on using the oculus sdk at all and will be 100% steamvr so having the oculus sdk tracking isn't beneficial. Editor again works flawlessly for both oculus and index, vive pro etc. all running in steamvr mode with steamvr input bindings fully functional both in packaging and editor. Tracking is simply broken in packaging. SteamVR is my only VR engine plugin enabled. I'm using VR hands solver, VRE, OpenVRE, SteamVR input, electric nodes and AnimIK plugins for this project.

**Comment by uno1982** [comment]
Rune, someone pointed out that this option needs to be setup in steamworks. 
![43855069-B509-481A-B2FE-FB6AC7AE15EE](https://user-images.githubusercontent.com/45907836/76776379-ac220780-6774-11ea-8a52-287314aaa843.jpeg)
Should I be publishing to steam to confirm tracking? I’m currently launching the exe from the windowsnoeditor folder after packaging to test and the result is tracking is broken. Should I be publishing to steam and launching from there? I’ve never had problems testing my packaged builds directly from the folder executable before.

**Comment by 1runeberg** [comment]
@uno1982 - I just had a look at your logs, is the game loading properly to a level? It seems to shutdown immediately on load, at least based on the logs.

There's also a number of plugins here that manipulates/affects the hands. For VRE, there used to be an Input module that also does tracking and SteamVRInput - can you check if this is disabled on your project? VRE is getting loaded first so if its handling input this can cause issues.

You also have a couple of other plugins all manipulating the hands, and not sure if a combination or any of them may be interfering with basic tracking. 

If possible, try on a test level and start disabling plugins one at a time to check for any incompatibilities.

If you can also test the sample project we provide here:
https://github.com/ValveSoftware/steamvr_unreal_plugin/wiki/sample/SteamVRInputPlugin.zip

And see if that works on your end. Then try enabling the plugins you have one at a time on it to see which one may be causing the incompatibility.

Re: Steamworks
Since you are running locally and not via Steam, you should be able to run your VR Game/Experience with tracking and input as-is. From the logs, Steam VR is also finding the correct action manifest to load.

**Comment by uno1982** [comment]
Rune, Its loading into the level fine... the log is probably appearing that way because I have a intro level that displays a short intro scene then as soon as it loaded up into the lobby map and I realized it didn't work I exited the game via the escape key. I'm downloading the example project you provided. I'll swap to my custom build and add in each plugin one by one until I find the issue.
![plugins](https://user-images.githubusercontent.com/45907836/76784817-52c0d500-6782-11ea-9eb8-6ca6235d0874.PNG)

**Comment by 1runeberg** [comment]
ok, the teardown was just one second from the first level booting and then all plugins were unloaded and the engine shutdown.... so maybe I got the wrong logs?

also @uno1982 - make sure your builds are 64-bit not 32-bit.

**Comment by uno1982** [comment]
I only package in 64 bit so no worries there. This project is currently on steam in closed alpha using the legacy input without issues. I’ll update you once I package out the example project and my findings as I add in plugins. I’ll package as is test then I’ll start adding once I confirm the base project works. I’ll start with VRE as the first plugin.

Test 1: Package without any changes in 4.22.3 source engine. - successful 

Test 2: Packaged with ElectricNodes and VRExpansion -successful

Test 3: Packaged with ElectricNodes, VRExpansion, OpenVRExpansion - Successful

Test 4: ElectricNodes, VRExpansion, OpenVRExpansion, AdvancedSessions - Successful

Test 5: ElectricNodes, VRExpansion, OpenVRExpansion, AdvancedSessions, AnimIK, AsyncSave, VRHandsolver, EpicLeaderboard - Successful

Looks like the example project doesn't have any issues when adding these plugins and packaging out. Tracking still works properly. Maybe referencing one of the modules nodes is causing issues? I'm going to grab a blank VRE template from Mords repo and add the input plugin to it and package out to see if that alone causes issues. If this is successful I have a feeling it’s the VRHandsolver reference in my character. I’m not sure that simply loading the module is causing issues but the node call in the character may be a problem. I’ll update you shortly.

**Comment by uno1982** [comment]
I contacted Mordentral after performing the above test and also seeing that the packaging lost tracking after only having the steamvr input plugin added to a blank VRE template. The issue is that I need to explore this more because I didn't call any classes within the steamvrinput plugin example project. The other thing I'd like to check is using a character instead of a simple pawn as used in your example. As soon as I have more info I'll update this issue.

**Comment by uno1982** [comment]
I was able to replace the pawn class in your example with the vre vrcharacter class and packaged without issue with tracking working. I’m thinking there is still a reference to steamvr somewhere in the VRE 4.22 locked template and my project where the plugin itself loaded into a fresh project gets around this issue

**Comment by uno1982** [comment]
Mordentral followed up thinking it’s related to issue #49

**Comment by uno1982** [comment]
https://1drv.ms/u/s!ArlMtU4-teLakCDEmDKDgvjJF7w8?e=a8q3TZ
Rune,
Here is the template project with the input plugin. Nothing really to it just setup to load into the steamvr hands map with same setup as your project. I also have your project with vre loaded into it using the vrcharacter class from vre that works without issue if you'd like me to link it as well. Seems there may be a project setting or config ini that is the issue? Mind debugging against this and pointing me in the right direction?  I feel like if the issue here is discovered my project will likely be suffering the same fate and can be corrected by the same resolve. Starting to believe it’s a ini, project setting or something along those lines as the steamvr hands map is using the default pawn and motioncontroller with gamemode base and shouldn’t be calling the vre module at all and still presents with tracking issues when packaged.

**Comment by uno1982** [comment]
I would like to state that this issue seems related to something with the VRETemplate and how the SteamVR input plugin works within that template at that version number. The above example shows that packaging the SteamVR Hands map (default pawn and motioncontroller) from the SteamVR Input plugin and launching results in tracking loss but the binds are all working and manifest files loaded within the logs properly. An interesting twist to this is that I can update the 4.22.3 projects to 4.24.3 without any changes and package to get around this issue. The result being that the input binding names change from the 4.22.3 version, motioncontroller is no longer available (due to being deprecated), Touch bindings no longer conflict with the Oculus SDK since the "STEAMVR inputs are gone" the oculus touch bindings seem to drive both sdk and steam now and tracking just works.

**Comment by 1runeberg** [comment]
HI @uno1982 

**Re: Issue #49** 
This has been fixed since last year, and the fix is incorporated in the release branches as well as the integrated version in 4.24

**Re: Upgrade**
There shouldve been some changes though for the upgrade, the key names wouldnt match exactly so they need to be rebound, just double check those if you haven't already to stave off any potential future issues.

**Re:VRETemplate**
I'm not sure if it was you or someone else who sent me the same VRETemplate. Not sure if it's project setting related as I would need to completely understand the VRE code, so this is a better question to post to the template author.

I did have a quick look at the standard settings and they appear to be alright. For potential debugging, I would look at the custom motion controller components the plugin has and potentially the network bits of it to see what's going on with tracking in builds.

This is outside the scope of our support though, as the plugin itself works as designed and follows strictly the interface paths provided by Epic for VR Platform providers/plugins. We can't make changes for individual plugins and risk other games breaking, a lot of titles including major (wide distribution/player base) ones have shipped with the Marketplace plugin already, unless of course it is a bug on our end but this doesn't seem to be the case here.

Sorry I don't have much good news for you here. But for the (community?) provided solution of migrating the template to a new project in the same engine version, the best tip I can provide is doing a migrate of the Levels instead of individual assets. Migrating Levels will have UE migrate all the assets that that level uses (make sure that your new project does have all the plugins first), this may make that chore a bit easier. 

You might also want to try retargetting the project to an engine (same version) from a source build to a binary (launcher) build and flipping between the two and see if that resuscitates the template somehow and may save you a migrate.
